tests/bluetooth/bsim_bt/bsim_test_l2cap_stress/src/main_l2cap_stress.c

```diff
@@ -0,0 +1,487 @@
+	/* LOG_ERR("%s ret %d", ret); */
+
+	if (ret < 0) {
+		FAIL("L2CAP error %d\n", ret);
+		net_buf_unref(buf);
+	}
+
+	LOG_WRN("sent %d len %d", ret, len);
+	return ret;
+}
+
+struct net_buf *alloc_buf_cb(struct bt_l2cap_chan *chan)
+{
+	return net_buf_alloc(&local_l2cap_rx_pool, K_NO_WAIT);
+}
+
+static int get_l2cap_chan(struct bt_l2cap_chan *chan) {
+	for (int i=0; i<NUM_L2CAP_CHANS; i++) {
+		if (l2cap_chans[i] == chan) return i;
+	}
+
+	FAIL("Channel %p not found\n", chan);
+	return -1;
+}
+
+static void register_channel(struct bt_l2cap_chan *chan)
+{
+	int i = get_l2cap_chan(NULL);
+
+	l2cap_chans[i] = chan;
+}
+
+void sent_cb(struct bt_l2cap_chan *chan)
+{
+	LOG_WRN("%s %p", __func__, chan);
+	int idx = get_l2cap_chan(chan);
+
+	if (tx_left[idx]) {
+		tx_left[idx]--;
+		l2cap_chan_send(chan, tx_data, sizeof(tx_data));
+	} else {
+		LOG_ERR("Done sending %p", chan->conn);
+	}
+}
+
+uint16_t rx_cnt = 0;
+int recv_cb(struct bt_l2cap_chan *chan, struct net_buf *buf)
+{
+	LOG_WRN("%s len %d", __func__, buf->len);
+	LOG_HEXDUMP_DBG(buf->data, buf->len, "RAW");
+	rx_cnt++;
+
+	return 0;
+}
+
+void l2cap_chan_connected_cb(struct bt_l2cap_chan *l2cap_chan)
+{
+	struct bt_l2cap_le_chan *chan =
+		CONTAINER_OF(l2cap_chan, struct bt_l2cap_le_chan, chan);
+
+	SET_FLAG(flag_l2cap_connected);
+	LOG_DBG("%x (tx mtu %d mps %d) (tx mtu %d mps %d)",
+		l2cap_chan,
+		chan->tx.mtu,
+		chan->tx.mps,
+		chan->rx.mtu,
+		chan->rx.mps);
+
+	register_channel(l2cap_chan);
+}
+
+void l2cap_chan_disconnected_cb(struct bt_l2cap_chan *chan)
+{
+	UNSET_FLAG(flag_l2cap_connected);
+	LOG_DBG("%x", chan);
+}
+
+static struct bt_l2cap_chan_ops ops = {
+	.connected = l2cap_chan_connected_cb,
+	.disconnected = l2cap_chan_disconnected_cb,
+	.alloc_buf = alloc_buf_cb,
+	.recv = recv_cb,
+	.sent = sent_cb,
+};
+
+#define HOST_MAX_L2CAP_CHANNELS 10
+
+static struct bt_l2cap_le_chan l2cap_channels[HOST_MAX_L2CAP_CHANNELS];
+
+struct bt_l2cap_le_chan *get_free_l2cap_le_chan(void)
+{
+	for (int i = 0; i < HOST_MAX_L2CAP_CHANNELS; i++) {
+		struct bt_l2cap_le_chan *le_chan = &l2cap_channels[i];
+
+		if (le_chan->state != BT_L2CAP_DISCONNECTED) {
+			continue;
+		}
+
+		memset(le_chan, 0, sizeof(*le_chan));
+		return le_chan;
+	}
+
+	return NULL;
+}
+
+int server_accept_cb(struct bt_conn *conn, struct bt_l2cap_chan **chan)
+{
+	struct bt_l2cap_le_chan *le_chan = NULL;
+
+	le_chan = get_free_l2cap_le_chan();
+	if (le_chan == NULL) {
+		return -ENOMEM;
+	}
+
+	memset(le_chan, 0, sizeof(*le_chan));
+	le_chan->chan.ops = &ops;
+	le_chan->rx.mtu = l2cap_mtu;
+	le_chan->rx.init_credits = INIT_CREDITS;
+	le_chan->tx.init_credits = INIT_CREDITS;
+	*chan = &le_chan->chan;
+
+	return 0;
+}
+
+static struct bt_l2cap_server test_l2cap_server = {
+	.accept = server_accept_cb
+};
+
+static int l2cap_server_register(bt_security_t sec_level)
+{
+	test_l2cap_server.psm = 0;
+	test_l2cap_server.sec_level = sec_level;
+
+	__ASSERT_NO_MSG(bt_l2cap_server_register(&test_l2cap_server) == 0);
```

Comment:
Don't put side effects in assert macros. :O

Reply:
Preempt: The day will come when we need to run babelsim tests with assertions disabled. And somebody may be tempted to copy this code into their application. I think it's reasonable that the babelsim tests double as samples in addition to tests.